### Import required dependencies

In [71]:
import pandas as pd
import os

## Deliverable 1: Collect the Data

To collect the data that you’ll need, complete the following steps:

1. Using the Pandas `read_csv` function and the `os` module, import the data from the `new_full_student_data.csv` file, and create a DataFrame called student_df. 

2. Use the head function to confirm that Pandas properly imported the data.


In [72]:
# Create the path and import the data
full_student_data = os.path.join('C:/Users/gopok/Downloads/new_full_student_data.csv')
student_df = pd.read_csv(full_student_data)

In [73]:
# Verify that the data was properly imported
student_df.head()

student_id      student_name grade           school_name  reading_score  \
0   103880842     Travis Martin   9th  Sullivan High School           59.0   
1    45069750     Michael Brown   9th     Dixon High School           94.7   
2    45024902   Gabriela Lucero   9th    Wagner High School           89.0   
3    62582498  Susan Richardson   9th     Silva High School           69.7   
4    16437227      Sherry Davis  11th    Bowers High School            NaN   

   math_score school_type  school_budget  
0        88.2      Public         961125  
1        73.5     Charter         870334  
2        70.4      Public         846745  
3        80.3      Public         991918  
4        27.5      Public         848324

## Deliverable 2: Prepare the Data

To prepare and clean your data for analysis, complete the following steps:
    
1. Check for and remove all rows with `NaN`, or missing, values in the student DataFrame. 

2. Check for and remove all duplicate rows in the student DataFrame.

3. Use the `str.replace` function to remove the "th" from the grade levels in the grade column.

4. Check data types using the `dtypes` property.

5. Remove the "th" suffix from every value in the grade column using `str` and `replace`.

6. Change the grade colum to the `int` type and verify column types.

7. Use the head (and/or the tail) function to preview the DataFrame.

In [74]:
# Check for null values
student_df.isnull().sum()

student_id          0
student_name        0
grade               0
school_name         0
reading_score    1968
math_score        982
school_type         0
school_budget       0
dtype: int64

In [75]:
# Drop rows with null values and verify removal
student_df=student_df.dropna(how='all',subset=['reading_score'])
student_df=student_df.dropna(how='all',subset=['math_score'])
student_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 16667 entries, 0 to 19513
Data columns (total 8 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   student_id     16667 non-null  int64  
 1   student_name   16667 non-null  object 
 2   grade          16667 non-null  object 
 3   school_name    16667 non-null  object 
 4   reading_score  16667 non-null  float64
 5   math_score     16667 non-null  float64
 6   school_type    16667 non-null  object 
 7   school_budget  16667 non-null  int64  
dtypes: float64(2), int64(2), object(4)
memory usage: 1.1+ MB


In [76]:
# Check for duplicated rows
student_df.duplicated().sum()

1836

In [77]:
# Drop duplicated rows and verify removal
student_df=student_df.drop_duplicates()
student_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 14831 entries, 0 to 19513
Data columns (total 8 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   student_id     14831 non-null  int64  
 1   student_name   14831 non-null  object 
 2   grade          14831 non-null  object 
 3   school_name    14831 non-null  object 
 4   reading_score  14831 non-null  float64
 5   math_score     14831 non-null  float64
 6   school_type    14831 non-null  object 
 7   school_budget  14831 non-null  int64  
dtypes: float64(2), int64(2), object(4)
memory usage: 1.0+ MB


In [78]:
# Check data types
student_df.dtypes

student_id         int64
student_name      object
grade             object
school_name       object
reading_score    float64
math_score       float64
school_type       object
school_budget      int64
dtype: object

In [79]:
# Examine the grade column to understand why it is not an int
student_df['grade']
# Error is happening due to "th" in the grade column as it is considered as string rather than numeric

0         9th
1         9th
2         9th
3         9th
5         9th
         ... 
19508    10th
19509    12th
19511    11th
19512    11th
19513    12th
Name: grade, Length: 14831, dtype: object

In [80]:
# Remove the non-numeric characters and verify the contents of the column
student_df['grade'] = student_df['grade'].str.replace(r'\D', '')
student_df.head()

student_id      student_name grade             school_name  reading_score  \
0   103880842     Travis Martin     9    Sullivan High School           59.0   
1    45069750     Michael Brown     9       Dixon High School           94.7   
2    45024902   Gabriela Lucero     9      Wagner High School           89.0   
3    62582498  Susan Richardson     9       Silva High School           69.7   
5    74579444   Cynthia Johnson     9  Montgomery High School           63.5   

   math_score school_type  school_budget  
0        88.2      Public         961125  
1        73.5     Charter         870334  
2        70.4      Public         846745  
3        80.3      Public         991918  
5        76.9     Charter         893368

In [81]:
# Change the grade column to the int type and verify column types
student_df['grade'] = student_df['grade'].str.replace(r'\D', '').astype(int)

## Deliverable 3: Summarize the Data

Describe the data using summary statistics on the data as a whole and on individual columns.

1. Generate the summary statistics for each DataFrame by using the `describe` function.

2. Display the mean math score using the `mean` function. 

2. Store the minimum reading score as `min_reading_score`.

In [82]:
# Display summary statistics for the DataFrame
student_df.describe()

student_id         grade  reading_score    math_score  school_budget
count  1.483100e+04  14831.000000   14831.000000  14831.000000   14831.000000
mean   6.975296e+07     10.355539      72.357865     64.675733  893742.749107
std    3.452909e+07      1.097728      15.224590     15.844093   53938.066467
min    1.000906e+07      9.000000      10.500000      3.700000  817615.000000
25%    3.984433e+07      9.000000      62.200000     54.500000  846745.000000
50%    6.965978e+07     10.000000      73.800000     65.300000  893368.000000
75%    9.927449e+07     11.000000      84.000000     76.000000  956438.000000
max    1.299997e+08     12.000000     100.000000    100.000000  991918.000000

In [83]:
# Display the mean math score using the mean function
student_df["math_score"].mean()

64.67573326141189

In [84]:
# Store the minimum reading score as min_reading_score
student_df["math_score"].min()

3.7

## Deliverable 4: Drill Down into the Data

Drill down to specific rows, columns, and subsets of the data.

To drill down into the data, complete the following steps:

1. Use `loc` to display the grade column.

2. Use `iloc` to display the first 3 rows and columns 3, 4, and 5.

3. Show the rows for grade nine using `loc`.

4. Store the row with the minimum overall reading score as `min_reading_row` using `loc` and the `min_reading_score` found in Deliverable 3.

5. Find the reading scores for the school and grade from the output of step three using `loc` with multiple conditional statements.

6. Using conditional statements and `loc` or `iloc`, find the mean reading score for all students in grades 11 and 12 combined.

In [85]:
# Use loc to display the grade column
student_df.loc[:,['grade']]

grade
0          9
1          9
2          9
3          9
5          9
...      ...
19508     10
19509     12
19511     11
19512     11
19513     12

[14831 rows x 1 columns]

In [86]:
# Use `iloc` to display the first 3 rows and columns 3, 4, and 5.
student_df.iloc[:3,[3,4,5]]

school_name  reading_score  math_score
0  Sullivan High School           59.0        88.2
1     Dixon High School           94.7        73.5
2    Wagner High School           89.0        70.4

In [87]:
# Select the rows for grade nine and display their summary statistics using `loc` and `describe`.
ninth_grader = student_df[student_df['grade']==9]
ninth_grader.describe()

student_id   grade  reading_score   math_score  school_budget
count  4.132000e+03  4132.0    4132.000000  4132.000000    4132.000000
mean   6.979441e+07     9.0      69.236713    66.585624  898692.606002
std    3.470565e+07     0.0      15.277354    16.661533   54891.596611
min    1.000906e+07     9.0      17.900000     5.300000  817615.000000
25%    3.953848e+07     9.0      59.000000    56.000000  846745.000000
50%    6.984037e+07     9.0      70.050000    67.800000  893368.000000
75%    9.939504e+07     9.0      80.500000    78.500000  957299.000000
max    1.299997e+08     9.0      99.900000   100.000000  991918.000000

In [88]:
# Store the row with the minimum overall reading score as `min_reading_row`
min_reading_row = student_df[student_df.reading_score == student_df.reading_score.min()]
print(min_reading_row)
# using `loc` and the `min_reading_score` found in Deliverable 3.

      student_id    student_name  grade        school_name  reading_score  \
3706    81758630  Matthew Thomas     10  Dixon High School           10.5   

      math_score school_type  school_budget  
3706        58.4     Charter         870334  


In [89]:
# Use loc with conditionals to select all reading scores from 10th graders at Dixon High School.
tenth_grader_Dixon = student_df.loc[(student_df['grade'] == 10) & (student_df['school_name'] == "Dixon High School")]
print(tenth_grader_Dixon)

       student_id    student_name  grade        school_name  reading_score  \
45      103118608    Ryan Charles     10  Dixon High School           71.1   
60       86894214  Jeremiah Smith     10  Dixon High School           59.5   
69       53464729      Troy Doyle     10  Dixon High School           88.6   
94       41212637   Matthew Lewis     10  Dixon High School           81.5   
100      31831251   Jonathan Bell     10  Dixon High School           95.3   
...           ...             ...    ...                ...            ...   
19283    89972942   Carol Sanford     10  Dixon High School           52.9   
19306    33033950     John Gibson     10  Dixon High School           58.0   
19344    55784164   Cynthia Doyle     10  Dixon High School           38.0   
19368    29598470  Michelle Moore     10  Dixon High School           84.4   
19445    54346011     David Henry     10  Dixon High School           43.9   

       math_score school_type  school_budget  
45           93.

In [90]:
# Find the mean reading score for all students in grades 11 and 12 combined.
mean_read = student_df.loc[(student_df['grade'] > 10)]
mean_read['reading_score'].mean()

74.90038089192188

## Deliverable 5: Make Comparisons Between District and Charter Schools

Compare district vs charter schools for budget, size, and scores.

Make comparisons within your data by completing the following steps:

1. Using the `groupby` and `mean` functions, look at the average reading and math scores per school type.

1. Using the `groupby` and `count` functions, find the total number of students at each school.

2. Using the `groupby` and `mean` functions, find the average budget per grade for each school type.

In [91]:
# Use groupby and mean to find the average reading and math scores for each school type.
student_df.groupby(['school_type']).count()

student_id  student_name  grade  school_name  reading_score  \
school_type                                                                
Charter            6711          6711   6711         6711           6711   
Public             8120          8120   8120         8120           8120   

             math_score  school_budget  
school_type                             
Charter            6711           6711  
Public             8120           8120

In [92]:
# Use the `groupby`, `count`, and `sort_values` functions to find the
# total number of students at each school and sort from most students to least students.
new = student_df.groupby(['school_name','school_type']).mean()
new.sort_values(by=['student_id'], ascending=False)

student_id      grade  reading_score  \
school_name            school_type                                           
Chang High School      Charter      7.311166e+07  10.076023      66.676023   
Fisher High School     Public       7.066683e+07  10.500000      71.683208   
Wagner High School     Public       7.039240e+07  10.482803      73.548929   
Dixon High School      Charter      7.025468e+07  10.184460      74.107707   
Silva High School      Public       7.022071e+07  10.334536      73.317042   
Montgomery High School Charter      6.990783e+07  10.740432      70.025466   
Green High School      Charter      6.989955e+07  10.035186      75.667823   
Campbell High School   Charter      6.989035e+07  10.476658      71.142260   
Turner High School     Public       6.986789e+07   9.829787      75.854019   
Richard High School    Charter      6.957271e+07  10.453721      67.968240   
Bowers High School     Public       6.928458e+07  11.114570      76.250062   
Odonnell High School   Public       6.915767e+07  10.836601      69.105011   
Campos High School     Public       6.872389e+07   9.818854      67.920887   
Woods High School      Public       6.828824e+07  10.487643      67.939924   
Sullivan High School   Public       6.819592e+07   9.878476      71.816993   

                                    math_score  school_budget  
school_name            school_type                             
Chang High School      Charter       62.412281       840132.0  
Fisher High School     Public        65.999123       817615.0  
Wagner High School     Public        64.962881       846745.0  
Dixon High School      Charter       69.105433       870334.0  
Silva High School      Public        60.692245       991918.0  
Montgomery High School Charter       66.812512       893368.0  
Green High School      Charter       66.224171       832670.0  
Campbell High School   Charter       62.179115       960726.0  
Turner High School     Public        66.044208       956438.0  
Richard High School    Charter       66.490381       889699.0  
Bowers High School     Public        53.189041       848324.0  
Odonnell High School   Public        67.793464       959474.0  
Campos High School     Public        59.357486       957299.0  
Woods High School      Public        66.093536       912243.0  
Sullivan High School   Public        61.523996       961125.0

# Deliverable 6: Summarize Your Findings
In the cell below, write a few sentences to describe any discoveries you made while performing your analysis along with any additional analysis you believe would be worthwhile.

*your summary here*
- There were initially 08 columns and 19514 rows in the data set. Post cleaning which includes removing rows having missing values in reading_score and math_score along with removing duplicate rows, at last we are left with 14,831 rows for the analysis.
- Average budget of the public schools are higher but the average reading score and maths score is higher for the charter schools
- Average reading score for the students in grade 11 and 12 combined is ~75
- Maximum score in math is a perfect 100 while in reading it is 99.9
- Minimum score in math is 5 while in reading it is better as compared to math
- Average reading score is 69 and math score is 66 irrespective of the school type
- 45% of students have enrolled themselves in charter schools and remainings are in public school
- Montgomery High School (Charter) has the highest number of students and Wagner High School has the highest number of student among all the public school
- Bowers High School (Public) has the highest overall mean reading score
- Dixon High School (Charter) has the highest overall mean math score